File: src/lib/clock.ts
// The simulation clock — the single source of "now" for the living world. Decoupled from wall-clock: it
// accumulates real frame dt (scaled by `rate`) into whole FIXED-SIZE ticks, so the sim always advances in
// deterministic integer steps. Canonical time IS the integer `tick` (seconds = tick × DT). Paired with the
// stateless RNG (rng.ts, keyed by tick) the whole world becomes a pure function of (seed, tick): you can
// pause, speed up / slow down, single-step, and SEEK to any tick — the basis for time travel.
//
// Determinism contract: identical TICK ⇒ identical world (because all randomness is rand(seed, tick, …)
// and the sim's per-tick update is a pure function of the previous tick). Two machines may sit at
// different ticks at the same wall-clock instant — frame pacing varies — but seeking both to tick T shows
// the SAME world. That's the right guarantee for a real-time sim, and it's what lets a share link say
// "this world at tick T" and have it reproduce.
//
// Plain (no runes) on purpose: it's read every frame in the sim hot path — UI mirrors clock.tick/.time
// into its own $state for display (same pattern as agentManager). See docs/self-sustaining-world.md.

export const DT = 1 / 30; // seconds per sim tick (fixed timestep)
const MAX_CATCHUP = 6; // cap ticks emitted per advance() so a long frame stall can't spiral the sim
const EPS = 1e-9; // float slack: a full tick's worth of accumulated dt should never be lost to rounding

export type TickListener = (tick: number) => void;
export type SeekListener = (tick: number, from: number) => void;

export class SimClock {
	tick = 0; // integer sim step — the canonical clock position
	rate = 1; // tick-speed multiplier (2 = double-time, 10 = time-lapse); 0 ≈ paused
	playing = true;
	#acc = 0; // leftover sub-tick real-time (seconds), carried between frames
	#onTick: TickListener[] = [];
	#onSeek: SeekListener[] = [];

	/** Seconds of simulated time elapsed (tick × DT). */
	get time(): number {
		return this.tick * DT;
	}

	/** Not advancing right now (stopped or rate 0). */
	get paused(): boolean {
		return !this.playing || this.rate === 0;
	}

	/** Feed real elapsed seconds each frame; emits whole, rate-scaled ticks to onTick listeners.
	 *  Returns how many ticks advanced this call. */
	advance(realDt: number): number {
		if (this.paused || !(realDt > 0)) return 0;
		this.#acc += realDt * this.rate;
		let n = 0;
		while (this.#acc >= DT - EPS && n < MAX_CATCHUP) {
			this.#acc -= DT;
			n++;
		}
		if (this.#acc >= DT - EPS) this.#acc = 0; // dropped backlog beyond the catch-up cap (doesn't affect
		// reproducibility AT a given tick — only how fast wall-clock maps to ticks during a stall)
		for (let i = 0; i < n; i++) {
			this.tick++;
			this.#emit(this.tick);
		}
		return n;
	}

	/** Advance exactly n ticks now (manual single-step while paused, or scripted replay/fast-forward). */
	step(n = 1): void {
		const k = Math.max(0, Math.floor(n));
		for (let i = 0; i < k; i++) {
			this.tick++;
			this.#emit(this.tick);
		}
	}

	/** TIME TRAVEL: jump to an absolute tick. Fires onSeek(target, from) so the sim can reconstruct state
	 *  for that moment (restore the nearest checkpoint ≤ target, then replay forward — deterministic,
	 *  because every tick is rand(seed, tick, …)). Does NOT emit per-tick; reconstruction is the sim's job. */
	seek(targetTick: number): void {
		const t = Math.max(0, Math.floor(targetTick));
		const from = this.tick;
		if (t === from) return;
		this.tick = t;
		this.#acc = 0;
		for (const cb of this.#onSeek) cb(t, from);
	}

	pause(): void {
		this.playing = false;
	}
	play(): void {
		this.playing = true;
	}
	toggle(): void {
		this.playing = !this.playing;
	}
	/** Set the tick-speed multiplier (clamped ≥ 0). */
	setRate(r: number): void {
		this.rate = Math.max(0, r);
	}
	/** Hard reset to a tick (default 0) with no replay signal — for loading a fresh/shared world. */
	reset(tick = 0): void {
		this.tick = Math.max(0, Math.floor(tick));
		this.#acc = 0;
	}

	/** Subscribe to each whole tick advance; returns an unsubscribe fn. */
	onTick(cb: TickListener): () => void {
		this.#onTick.push(cb);
		return () => {
			const i = this.#onTick.indexOf(cb);
			if (i >= 0) this.#onTick.splice(i, 1);
		};
	}

	/** Subscribe to seeks (time-travel jumps); returns an unsubscribe fn. */
	onSeek(cb: SeekListener): () => void {
		this.#onSeek.push(cb);
		return () => {
			const i = this.#onSeek.indexOf(cb);
			if (i >= 0) this.#onSeek.splice(i, 1);
		};
	}

	#emit(tick: number): void {
		for (const cb of this.#onTick) cb(tick);
	}
}

/** App-wide simulation clock. */
export const clock = new SimClock();

File: crates/worldsim/src/clock.rs
//! Fixed-timestep simulation clock — port of `src/lib/clock.ts` (the time-accounting half:
//! advance / step / seek / alpha). Listeners (`onTick`/`onSeek`) are a VIEW concern and stay in JS — the
//! Rust core's step loop calls the sim tick directly. Canonical time is the integer `tick` (seconds =
//! tick × DT). Pure deterministic f64 accounting → identical results across browser / worker / replay; the
//! same arithmetic as the JS source (verified in tests against a captured reference sequence).

pub const DT: f64 = 1.0 / 30.0; // seconds per sim tick (fixed timestep)
const MAX_CATCHUP: u32 = 6; // cap ticks per advance() so a long frame stall can't spiral the sim
const EPS: f64 = 1e-9; // float slack so a full tick's worth of dt isn't lost to rounding

#[derive(Clone, Copy, Debug)]
pub struct SimClock {
    pub tick: i64,    // integer sim step — the canonical clock position
    pub rate: f64,    // tick-speed multiplier (2 = double-time, 0 ≈ paused)
    pub playing: bool,
    acc: f64,         // leftover sub-tick real-time (seconds), carried between frames
}

impl Default for SimClock {
    fn default() -> Self {
        Self { tick: 0, rate: 1.0, playing: true, acc: 0.0 }
    }
}

impl SimClock {
    pub fn new() -> Self {
        Self::default()
    }

    /// Simulated seconds elapsed (tick × DT).
    pub fn time(&self) -> f64 {
        self.tick as f64 * DT
    }

    /// Not advancing right now (stopped or rate 0).
    pub fn paused(&self) -> bool {
        !self.playing || self.rate == 0.0
    }

    /// Sub-tick interpolation factor (0..1) — how far real time has advanced toward the next tick. The
    /// render side lerps between the previous and current sim step by this.
    pub fn alpha(&self) -> f64 {
        let a = self.acc / DT;
        if a < 0.0 {
            0.0
        } else if a > 1.0 {
            1.0
        } else {
            a
        }
    }

    /// Feed real elapsed seconds; returns how many whole, rate-scaled ticks advanced (the caller runs the
    /// sim that many steps). Capped at MAX_CATCHUP so a stall can't spiral; backlog beyond the cap is dropped
    /// (doesn't affect reproducibility AT a given tick, only how fast wall-clock maps to ticks during a stall).
    pub fn advance(&mut self, real_dt: f64) -> u32 {
        if self.paused() || !(real_dt > 0.0) {
            return 0;
        }
        self.acc += real_dt * self.rate;
        let mut n = 0u32;
        while self.acc >= DT - EPS && n < MAX_CATCHUP {
            self.acc -= DT;
            n += 1;
        }
        if self.acc >= DT - EPS {
            self.acc = 0.0;
        }
        self.tick += n as i64;
        n
    }

    /// Advance exactly n ticks now (manual single-step while paused, or scripted replay/fast-forward).
    pub fn step(&mut self, n: i64) {
        if n > 0 {
            self.tick += n;
        }
    }

    /// TIME TRAVEL: jump to an absolute tick (the caller reconstructs state for that moment).
    pub fn seek(&mut self, target_tick: i64) {
        self.tick = target_tick.max(0);
        self.acc = 0.0;
    }

    /// Hard reset to a tick (default 0) with no replay signal — for loading a fresh/shared world.
    pub fn reset(&mut self, tick: i64) {
        self.tick = tick.max(0);
        self.acc = 0.0;
    }

    /// Set the tick-speed multiplier (clamped ≥ 0).
    pub fn set_rate(&mut self, r: f64) {
        self.rate = r.max(0.0);
    }

    pub fn pause(&mut self) {
        self.playing = false;
    }
    pub fn play(&mut self) {
        self.playing = true;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Reference sequence captured from src/lib/clock.ts — the (n_ticks, tick, alpha) at each step MUST match.
    #[test]
    fn advance_parity() {
        let mut c = SimClock::new();

        assert_eq!(c.advance(DT * 2.5), 2); // 2 whole ticks…
        assert_eq!(c.tick, 2);
        assert!((c.alpha() - 0.5).abs() < 1e-9); // …0.5 carried

        assert_eq!(c.advance(DT * 0.5), 1); // carried 0.5 + 0.5 = 1 tick
        assert_eq!(c.tick, 3);
        assert_eq!(c.alpha(), 0.0);

        assert_eq!(c.advance(DT * 10.0), 6); // a big stall → capped at MAX_CATCHUP, backlog dropped
        assert_eq!(c.tick, 9);
        assert_eq!(c.alpha(), 0.0);

        c.rate = 2.0;
        assert_eq!(c.advance(DT), 2); // double-time
        assert_eq!(c.tick, 11);

        c.playing = false;
        assert_eq!(c.advance(DT * 5.0), 0); // paused → no advance
        assert_eq!(c.tick, 11);
    }

    #[test]
    fn seek_reset_clamp() {
        let mut c = SimClock::new();
        c.advance(DT * 3.0);
        c.seek(100);
        assert_eq!(c.tick, 100);
        assert_eq!(c.alpha(), 0.0); // seek clears the sub-tick accumulator
        c.seek(-5);
        assert_eq!(c.tick, 0); // clamped ≥ 0
        c.reset(50);
        assert_eq!(c.tick, 50);
        c.set_rate(-3.0);
        assert_eq!(c.rate, 0.0); // clamped ≥ 0
    }
}
